nym-sphinx-wasm/src/lib.rs

```diff
@@ -0,0 +1,142 @@
+pub struct NodeData {
+    address: String,
+    public_key: String,
+}
+/// Creates a Sphinx packet for use in JavaScript applications, using wasm.
+///
+/// The `wasm-pack build` command will cause this to output JS bindings and a
+/// wasm executable in the `pkg/` directory.
+///
+/// Message chunking is currently not implemented. If the message exceeds the
+/// capacity of a single Sphinx packet, the extra information will be discarded.
+#[wasm_bindgen]
+pub fn create_sphinx_packet(raw_route: &str, msg: &str, destination: &str) -> Vec<u8> {
+    utils::set_panic_hook(); // nicer js errors.
+
+    let route = sphinx_route_from(raw_route);
+
+    let average_delay = Duration::from_secs_f64(0.1);
+    let delays = delays::generate_from_average_duration(route.len(), average_delay);
+    let dest_bytes = DestinationAddressBytes::from_base58_string(destination.to_owned());
+    let dest = Destination::new(dest_bytes, [4u8; IDENTIFIER_LENGTH]);
+    let message = split_and_prepare_payloads(&msg.as_bytes()).pop().unwrap();
+    let sphinx_packet = match SphinxPacket::new(message, &route, &dest, &delays).unwrap() {
+        SphinxPacket { header, payload } => SphinxPacket { header, payload },
+    };
+
+    payload(sphinx_packet, route)
+}
+
+/// Concatenate the first mix address bytes with the Sphinx packet.
+///
+/// The Nym gateway node has no idea what is inside the Sphinx packet, or where
+/// it should send a packet it receives. So we prepend the packet with the
+/// address bytes of the first mix inside the packet, so that the gateway can
+/// forward the packet to it.
+fn payload(sphinx_packet: SphinxPacket, route: Vec<Node>) -> Vec<u8> {
+    let mut packet = sphinx_packet.to_bytes();
+    let mut first_mix_address = route.first().unwrap().clone().address.to_bytes().to_vec();
+    let mut bytes: Vec<u8> = vec![];
+    bytes.append(&mut first_mix_address);
+    bytes.append(&mut packet);
+    bytes
+}
+
+/// Attempts to create a Sphinx route, which is a `Vec<sphinx::Node>`, from a
+/// JSON string.
+///
+/// # Panics
+///
+/// This function panics if the supplied `raw_route` json string can't be
+/// extracted to a `JsonRoute`.
+///
+/// # Panics
+///
+/// This function panics if `JsonRoute.nodes` doesn't contain at least 1
+/// node.
+///
+fn sphinx_route_from(raw_route: &str) -> Vec<Node> {
+    let json_route: JsonRoute = serde_json::from_str(raw_route).unwrap();
+
+    assert!(
+        json_route.nodes.len() > 0,
+        "Sphinx packet must route to at least one mixnode."
+    );
+
+    let mut sphinx_route: Vec<Node> = vec![];
+    for node_data in json_route.nodes.iter() {
+        let x = Node::try_from(node_data.clone()).expect("Malformed NodeData");
+        sphinx_route.push(x);
+    }
+    sphinx_route
+}
+
+// Converts a base58 &str into a public key (MontgomeryPoint)
+//
+// A TryInto would be nice here but this crate doesn't own MontgomeryPoint.
+fn public_key_from(s: &str) -> MontgomeryPoint {
```

Comment:
I forget, but what was the exact reason we're operating directly on MontgomeryPoints again? 
Also we could/should have `into` or `try_into` on `MixIndentityPublicKey`

Reply:
Yeah I'd rather not, if I can avoid it (same as the Sphinx thing I just mentioned in KeyBase). Is there an easy way we can avoid it? I do need a public key here, can I use `MixIdentityKey`, or would a type alias somewhere in `nymsphinx` help?

Reply:
I've mentioned this one in https://github.com/nymtech/nym/issues/184 so we don't miss it. Save it for a rainy day etc